### PR TITLE
mds: warn when clients are not advancing their oldest_client_tid

### PR DESCRIPTION
**Pull request by ukernel**

### CODE REVIEW

**Comment by ukernel** [comment]
please merge https://github.com/ceph/ceph/pull/4218 first

**Comment by ukernel** [comment]
http://pulpito.ceph.com/ubuntu-2015-04-19_22:39:04-fs-wip-yan-fs-testing---basic-multi/

File: src/mds/SessionMap.cc [hunk 1]
```diff
@@ -681,6 +681,8 @@ void Session::decode(bufferlist::iterator &p)
 {
   info.decode(p);
 
+  num_completed_requests = info.completed_requests.size();
```
Comment (gregsfortytwo) [inline src/mds/SessionMap.cc]: size() is a constant-time function on standard maps, and num_completed_requests looks to just be an identical value. Please remove it and use size() instead.

File: src/common/config_opts.h [hunk 2]
```diff
@@ -450,6 +450,8 @@ OPTION(mds_snap_min_uid, OPT_U32, 0) // The minimum UID required to create a sna
 OPTION(mds_snap_max_uid, OPT_U32, 65536) // The maximum UID allowed to create a snapshot
 OPTION(mds_snap_rstat, OPT_BOOL, false) // enable/disbale nested stat for snapshot
 OPTION(mds_verify_backtrace, OPT_U32, 1)
+// detect clients which aren't trimming completed requests
+OPTION(mds_max_completed_requests, OPT_U32, 100000)
```
Comment (gregsfortytwo) [inline src/common/config_opts.h]: This is a very large value, how'd you come up with it?

**Comment by gregsfortytwo** [comment]
This should have a test to make sure the warnings actually happen; otherwise we'll never notice if it breaks.

File: src/common/config_opts.h (hunk 2 above)
Comment (ukernel) [inline src/common/config_opts.h]: On Tue, Apr 21, 2015 at 2:35 AM, Gregory Farnum <notifications@github.com>
wrote:

> In src/common/config_opts.h
> <https://github.com/ceph/ceph/pull/4386#discussion_r28715940>:
>
> > @@ -450,6 +450,8 @@ OPTION(mds_snap_min_uid, OPT_U32, 0) // The minimum UID required to create a sna
> >  OPTION(mds_snap_max_uid, OPT_U32, 65536) // The maximum UID allowed to create a snapshot
> >  OPTION(mds_snap_rstat, OPT_BOOL, false) // enable/disbale nested stat for snapshot
> >  OPTION(mds_verify_backtrace, OPT_U32, 1)
> > +// detect clients which aren't trimming completed requests
> > +OPTION(mds_max_completed_requests, OPT_U32, 100000)
>
> This is a very large value, how'd you come up with it?
>

not so large I have a request blocked for > 30 seconds

**Comment by ukernel** [comment]
https://github.com/ceph/ceph-qa-suite/pull/417

**Pull request merged**